src/shared/credentialsManager.ts

```diff
@@ -49,13 +49,19 @@ export class CredentialsManager {
      * @param profileName Profile to set up
      */
     private async createCredentials(profileName: string): Promise<AWS.Credentials> {
-        const credentials: AWS.Credentials = new AWS.SharedIniFileCredentials({
+        let credentials: AWS.Credentials = new AWS.SharedIniFileCredentials({
             profile: profileName,
             tokenCodeFn: async (mfaSerial, callback) =>
                 await CredentialsManager.getMfaTokenFromUser(mfaSerial, profileName, callback)
         })
 
-        await credentials.getPromise()
+        try {
+            await credentials.getPromise()
+        } catch {
```

Comment:
Is there a more intelligent way that we can detect if this profile is a credential_process profile? The problem here is if regular credential loading fails - we'll lose the actual exception message because we'll attempt to fall-back to credential_process in all cases. Might lead to funny exception messages.

Reply:
`DefaultCredentialsFileReaderWriter` could gain a method that accepts a profile name and returns an enum value indicating the "type" of the profile. The "type" would be based on the presence of certain keys in the Profile. From this type we can load credential appropriately.

Reply:
I agree error handling isn't very good here. When I was debugging, the code was triggering [userCredentialsUtils.ts#addUserDataToContext](https://github.com/aws/aws-toolkit-vscode/blob/master/src/shared/credentials/userCredentialsUtils.ts#L188) which was swallowing the error. Maybe it's worth creating a separate bug to add a log message for errors there?

There are, however, other places that call `CredentialsManager.getCredentials` that could benefit from proper error handling here. I can think of two semi-intelligent things to do:
1. Read the `~/.aws/config` file to decide if profle is `ProcessCredentials` or `SharedIniFileCredentials`. I could try to re-use [credentialsFile.ts](https://github.com/aws/aws-toolkit-vscode/blob/master/src/shared/credentials/credentialsFile.ts#L48) for this.
2. Combine error messages in some meaningful way. eg. "Failed to get credentials for {0}. Tried ini file credentials with result {1}. Tried credentials_process with result {2}.". 

My preference is something like (2) because I think it opens the door for more credential types (see subclasses of [AWS credentials](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Credentials.html)).

Reply:
I actually think 1 is the cleaner approach to Chris' point, we're already reading the credential file to "discover" profiles, at that point we should be able to understand what type of credential it is. I'm actually surprised that the JS SDK doesn't handle this for us.

Reply:
@kiiadi -- interesting! AWS SDK has something called a [CredentialProviderChain](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/CredentialProviderChain.html#defaultProviders-property). To use the `defaultProviders`, we'd have to set `process.env.AWS_PROFILE` **before loading the SDK** so I think that's a no-go ([docs](https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/loading-node-credentials-shared.html)). We could create a custom credential provider chain which is simple:

```
    private async createCredentials(profileName: string): Promise<AWS.Credentials> {
        const provider = new AWS.CredentialProviderChain([
            () => new AWS.ProcessCredentials({ profile: profileName }),
            () => new AWS.SharedIniFileCredentials({
                profile: profileName,
                tokenCodeFn: async (mfaSerial, callback) =>
                    await CredentialsManager.getMfaTokenFromUser(mfaSerial, profileName, callback)
            }),
        ])
        return provider.resolvePromise();
    }
```

The order of the credential provider functions determines the error message displayed so keeping them in this order preserves the current behaviour.

So... in light of the new information about `CredentialProviderChain`. Do you guys still prefer we write our own "credential factory" based on ini file (option 1) or go with option 2 based on `CredentialProviderChain`?

Reply:
As we provide support for more credential types, I wonder two things:
* if a reasonable provider chain order can be established (are there any cases where one type must be tried first?)
* if it is reasonable to allow profile resolutions against an account to fail in trying to find a valid credential type

Reply:
> if a reasonable provider chain order can be established...
I'm guessing the SDK would recommend using [CredentialProviderChain.defaultProviders](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/CredentialProviderChain.html#defaultProviders-property) out of the box and if not, it's up to the application to decide a reasonable order. Unfortunately, I don't think aws-toolkit-vscode can use `defaultProviders` out of the box because it requires manipulating environment variables **before loading the SDK** ([doc](https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/loading-node-credentials-shared.html)).

> if it is reasonable to allow profile resolutions against an account to fail ...
My opinion is "yes". This is how `CredentialProviderChain.defaultProviders` works. For reference, the failures are pretty fast because these credential providers will throw if the `~/.aws/config` file does not contain enough information. For example, the `ProcessCredentials` promise rejects immediately if [no `credentials_process` is configured](https://github.com/aws/aws-sdk-js/blob/master/lib/credentials/process_credentials.js#L85) and similarly for `SharedIniFileCredentials` if there is [no access key or secret key](https://github.com/aws/aws-sdk-js/blob/master/lib/credentials/shared_ini_file_credentials.js#L142-L147).

Reply:
Sounds good to me. I also had a chat with someone from the JS SDK team, and the CredentialProviderChain seems appropriate. I'm happy with the CredentialProviderChain approach for this.

Reply:
Done